packages/plugin-compiler/src/main/kotlin/io/realm/compiler/IrUtils.kt

```diff
@@ -107,8 +109,20 @@ fun IrPluginContext.blockBody(
 val ClassDescriptor.isRealmObjectCompanion
     get() = isCompanionObject && (containingDeclaration as ClassDescriptor).hasRealmModelInterface
 
-val ClassDescriptor.hasRealmModelInterface
-    get() = getSuperInterfaces().firstOrNull { it.fqNameSafe == FqNames.REALM_MODEL_INTERFACE } != null
+val ClassDescriptor.hasRealmModelInterface: Boolean
+    get() {
+        // Using PSI to find super types to avoid cyclic reference (see https://github.com/realm/realm-kotlin/issues/339)
+        var hasRealmObjectAsSuperType = false
+        this.findPsi()?.acceptChildren(object : PsiElementVisitor() {
+            override fun visitElement(element: PsiElement) {
+                if (element.node.elementType == SUPER_TYPE_LIST) {
+                    hasRealmObjectAsSuperType = element.node.text.findAnyOf(listOf("RealmObject", "io.realm.RealmObject")) != null
```

Comment:
I thought about this use case, if needed we can also traverse the parent PSI to check that the user has imported the fully qualified name `io.realm.RealmObject` or `io.realm.*` when he uses the non fully qualified name `RealmObject`. This could be done by looking at the imports in the PSI 
```Kotlin
element.parent.parent.acceptChildren(object: PsiElementVisitor() {
                            override fun visitElement(element: PsiElement) {
                                when (element.node.elementType) {
                                    IMPORT_ALIAS,
                                    IMPORT_DIRECTIVE,
                                    IMPORT_LIST -> {
                                        hasImportedOurMarker = element.node.text.contains("io.realm.RealmObject")
                                    }
                                }
                            }
                        })
```
But at this stage, it's in the same probably as someone using our internal prefix `io_realm_kotlin_` which will clash which his variable name 🤷‍♂️

Reply:
I don't think this will cover all the edge cases as stated by [Jetbrains](https://kotlinlang.slack.com/archives/C7L3JB43G/p1653481272793489?thread_ts=1626430519.160400&cid=C7L3JB43G) and this will no longer be an issue on K2 compiler.